File: Python3/프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
import heapq

def solution(jobs):
    #작업 번호, 작업 요청 시각, 작업 소요 시간
    #작업을 안하고, 대기 큐가 비어있지 않다면, 가장 우선순위가 높은 작업을 대기큐에서 꺼내서 작업 진행
    #우선순위 - 작업 소요 시간이 짧고, 작업 요청시각이 빠르고, 작업 번호가 작은 것
    #작업 마치고, 요청 시점이 겹친다면, 요청이 들어온 작업을 대기 큐에 저장 후, 우선순위가 높은 작업을 대기 큐에서 꺼내서 시킴
    q = []
    dict = {}
    heapq.heapify(q)
    sec = 0
    block = 0
    answer = 0
    
    for i, item in enumerate(jobs):
        if item[0] in dict:
            dict[item[0]].append((item[1], item[0], i))
        else:
            dict[item[0]] = [(item[1], item[0], i)]
    
    while dict or q:
        if block > 0:
            block -= 1
        
        if sec in dict: #현재 시각에 들어오는 작업이 있을 경우
            for ls in dict[sec]:
                heapq.heappush(q, ls)
            dict.pop(sec)
            
        if block == 0 and q:
            current = heapq.heappop(q) #현재 가장 우선순위가 높은 작업 pop
            block = current[0]
            answer += (sec - current[1]) + block
                
        sec += 1
        
    
    answer //= len(jobs)
    
    return answer

File: C++/프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.cpp
#include <string>
#include <vector>
#include <queue>
#include <tuple>
#include <unordered_map>

using namespace std;
//소요시간이 짧은 것, 작업의 요청 시각이 빠른 것, 작업의 번호가 작은 것
priority_queue<tuple<int, int, int>> heap;
unordered_map<int, vector<int>> dic;

int solution(vector<vector<int>> jobs) {
    int answer = 0;
    int size = jobs.size();
    for (int i = 0; i < size; i++) {
        dic[jobs[i][0]].push_back(i); //작업 번호 넣기
    }
    int sec = 0;
    int curr = 0;
    
    while (!dic.empty() or !heap.empty()) {
        if (curr > 0) {
            curr--;
        }
        
        if (dic.find(sec) != dic.end()) {
            int dic_size = dic[sec].size();
            for (int i = 0; i < dic_size; i++) {
                int v = dic[sec][i];
                heap.push({-jobs[v][1], -sec, -v});
            }
            dic.erase(sec);
        }
        
        if (curr == 0 && !heap.empty()) {
             auto [a, b, c] = heap.top();
            heap.pop();
            answer += (sec -a + b);
            curr = -a;
        }
        
        sec++;  
        
    }
    answer = answer / size;
    return answer;
}